services/route_manager/src/main/java/com/futurewei/alcor/route/controller/NeutronRouterController.java

```diff
@@ -0,0 +1,363 @@
+            method = GET,
+            value = {"/project/{projectid}/routers"})
+    @DurationStatistics
+    public NeutronRoutersWebJson getNeutronRouters(@PathVariable String projectid) throws Exception {
+
+        List<NeutronRouterWebRequestObject> neutronRouters = new ArrayList<>();
+
+        Map<String, Router> routers = null;
+        RouterExtraAttribute routerExtraAttribute = null;
+
+        Map<String, Object[]> queryParams =
+                ControllerUtil.transformUrlPathParams(request.getParameterMap(), Router.class);
+
+        ControllerUtil.handleUserRoles(request.getHeader(ControllerUtil.TOKEN_INFO_HEADER), queryParams);
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            routers = this.routerDatabaseService.getAllRouters(queryParams);
+            if (routers == null) {
+                return new NeutronRoutersWebJson();
+            }
+
+            for (Map.Entry<String, Router> entry : routers.entrySet()) {
+                NeutronRouterWebRequestObject neutronRouterWebRequestObject = new NeutronRouterWebRequestObject();
+                Router router = (Router) entry.getValue();
+                routerExtraAttribute = this.routerExtraAttributeDatabaseService.getByRouterExtraAttributeId(router.getRouterExtraAttributeId());
+
+                BeanUtils.copyProperties(router, neutronRouterWebRequestObject);
+                if (routerExtraAttribute != null) {
+                    BeanUtils.copyProperties(routerExtraAttribute, neutronRouterWebRequestObject);
+                }
+
+                neutronRouters.add(neutronRouterWebRequestObject);
+            }
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw new Exception(e);
+        } catch (ResourceNotFoundException e) {
+            throw new Exception(e);
+        }
+
+        return new NeutronRoutersWebJson(neutronRouters);
+    }
+
+    /**
+     * Create a Neutron router
+     * @param projectid
+     * @param resource
+     * @return
+     * @throws Exception
+     */
+    @RequestMapping(
+            method = POST,
+            value = {"/project/{projectid}/routers"})
+    @DurationStatistics
+    public NeutronRouterWebJson createNeutronRouters(@PathVariable String projectid, @RequestBody NeutronRouterWebJson resource) throws Exception {
+
+        NeutronRouterWebRequestObject neutronRouterWebRequestObject = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+
+            // check resource
+            if (!RouteManagerUtil.checkNeutronRouterWebResourceIsValid(resource)) {
+                throw new ResourceNotValidException("request resource is invalid");
+            }
+
+            neutronRouterWebRequestObject = resource.getRouter();
+            String id = neutronRouterWebRequestObject.getId();
+
+            if (id == null || StringUtils.isEmpty(id)) {
```

Comment:
Could we consolidate the validation into one?

Reply:
we may do this in next PR